runtime/src/iree/hal/drivers/hip/hip_allocator.c

```diff
@@ -692,14 +692,14 @@ iree_status_t iree_hal_hip_allocator_free_async(
     return iree_ok_status();
   }
 
-  IREE_RETURN_IF_ERROR(IREE_HIP_CALL_TO_STATUS(allocator->symbols,
-                                               hipFree(device_ptr), "hipFree"));
   iree_hal_hip_buffer_set_allocation_empty(buffer);
 
   IREE_TRACE_FREE_NAMED(IREE_HAL_HIP_ALLOCATOR_ID, (void*)device_ptr);
   IREE_STATISTICS(iree_hal_allocator_statistics_record_free(
       &allocator->statistics, iree_hal_buffer_memory_type(buffer),
       iree_hal_buffer_allocation_size(buffer)));
+  IREE_RETURN_IF_ERROR(IREE_HIP_CALL_TO_STATUS(allocator->symbols,
+                                               hipFree(device_ptr), "hipFree"));
   return iree_ok_status();
```

Comment:
is the race with set_allocation_empty or with statistics tracing? if not with stats/tracing then it'd be good to keep it up with the non-debugging-logic - if you want it here then add some whitespace around it so it doesn't get lost in what looks like debugging code

```suggestion
      &allocator->statistics, iree_hal_buffer_memory_type(buffer),
      iree_hal_buffer_allocation_size(buffer)));
      
  IREE_RETURN_IF_ERROR(IREE_HIP_CALL_TO_STATUS(allocator->symbols,
                                               hipFree(device_ptr), "hipFree"));
                                               
  return iree_ok_status();
```

Reply:
The race is actually with the tracing code (specifically IREE_TRACE_FREE*) I will add the whitespace as appropriate. I can bring the "set_allocation_empty" along for the ride to keep it together.